csrc/preseg_passes/move_split_cat.cpp

```diff
@@ -58,79 +130,90 @@ class CancelSplitCat {
-
-  // Check all Exprs in `frontier`
-  // 1. have the same op type and attributes,
-  // 2. transform IDs in the same way, and
-  // 3. don't resize the split axis.
-
-  if (std::adjacent_find(
-          frontier.begin(), frontier.end(), [](Expr* lhs, Expr* rhs) {
-            return !lhs->sameOp(rhs);
-          }) != frontier.end()) {
-    return false;
-  }
+bool sameOp(const std::vector<Expr*>& frontier) {
+  return std::adjacent_find(
+             frontier.begin(), frontier.end(), [](Expr* lhs, Expr* rhs) {
+               return !lhs->sameOp(rhs);
+             }) == frontier.end();
+}
 
-  if (auto* set = dynamic_cast<LoadStoreOp*>(frontier[0])) {
-    if (set->opType() == LoadStoreOpType::Set) {
-      auto* set_out = set->out()->as<TensorView>();
-      std::optional<std::vector<int64_t>> permutation =
-          ir_utils::computePermutation(
-              set_out->getRootDomain(), set_out->getMaybeRFactorDomain());
-      if (!permutation.has_value()) {
+bool CancelSplitCat::sameIterDomainTransforms(
+    const std::vector<SliceOp*>& slices,
+    const std::vector<PadOp*>& pads) {
+  NVF_ERROR(slices.size() == pads.size());
+  NVF_ERROR(!slices.empty());
+
+  ValGraph& exact_graph = id_model_for_merging_.idGraph(IdMappingMode::EXACT);
+  {
+    const std::vector<IterDomain*>& first_rfactor =
+        slices[0]->out()->getMaybeRFactorDomain();
+    size_t num_dims = first_rfactor.size();
+    for (size_t i = 1; i < slices.size(); i++) {
+      const std::vector<IterDomain*>& rfactor =
+          slices[i]->out()->getMaybeRFactorDomain();
+      if (rfactor.size() != num_dims) {
         return false;
       }
+      for (size_t j = 0; j < num_dims; j++) {
+        exact_graph.mapVals(first_rfactor[j], rfactor[j]);
+      }
+    }
+  }
 
-      for (size_t i = 1; i < frontier.size(); i++) {
-        auto* other_set_out =
-            frontier[i]->as<LoadStoreOp>()->out()->as<TensorView>();
-        std::optional<std::vector<int64_t>> other_permutation =
-            ir_utils::computePermutation(
-                other_set_out->getRootDomain(),
-                other_set_out->getMaybeRFactorDomain());
-        if (!other_permutation.has_value()) {
-          return false;
-        }
-        if (*permutation != *other_permutation) {
+  for (PadOp* pad : pads) {
+    auto* pad_out = pad->out()->as<TensorView>();
+    if (id_model_for_merging_.hasSelfMapping(pad_out)) {
+      return false;
+    }
+  }
+
+  {
+    const std::vector<IterDomain*>& first_root =
+        pads[0]->out()->as<TensorView>()->getRootDomain();
+    size_t num_dims = first_root.size();
+    for (size_t i = 1; i < pads.size(); i++) {
+      const std::vector<IterDomain*>& root =
+          pads[i]->out()->as<TensorView>()->getRootDomain();
+      if (root.size() != num_dims) {
+        return false;
+      }
+      for (size_t j = 0; j < num_dims; j++) {
+        if (!exact_graph.disjointValSets().strictAreMapped(
+                first_root[j], root[j])) {
```

Comment:
it's likely just me being paranoid. But I'm not seeing logic where we are detecting the *order* of slice/cat being consistent. Can we add a test with
```
t0 = slice(input)
t1 = slice(input)
out = cat(t1, t0) # reversed order
```

Reply:
Done.